src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs

```diff
@@ -429,19 +433,21 @@ internal Location CreateLocation(AnnotatedCodeLocationVersionOne v1AnnotatedCode
                 Message = CreateMessage(message)
             };
 
-            LogicalLocation logicalLocation = new LogicalLocation();
+            logicalLocationKey = logicalLocationKey ?? fullyQualifiedLogicalName;
+
+            // Retrieve logical location so that we can acquire the index
+            LogicalLocation logicalLocation;
+            _v1KeyToV2LogicalLocationMap.TryGetValue(logicalLocationKey, out logicalLocation);            
 
-            logicalLocation.FullyQualifiedName = fullyQualifiedLogicalName;
-            logicalLocation.Name = GetLogicalLocationName(fullyQualifiedLogicalName);
-            location.FullyQualifiedLogicalName = logicalLocationKey;
+            location.FullyQualifiedLogicalName = fullyQualifiedLogicalName ?? logicalLocation?.FullyQualifiedName;
 
-            if (_logicalLocationToIndexMap.TryGetValue(logicalLocation, out int index))
+            int logicalLocationIndex;
+            if (logicalLocation == null || !_v2LogicalLocationToIndexMap.TryGetValue(logicalLocation, out logicalLocationIndex))
             {
-                index = _logicalLocationToIndexMap.Count;
-                _logicalLocationToIndexMap[logicalLocation] = index;
+                logicalLocationIndex = -1;
             }
```

Comment:
>} [](start = 12, length = 1)

This is right but I wonder if this is clearer:
```
int logicalLocationIndex = -1;
if (logicalLocation != null && _v2LogicalLocationToIndexMap.TryGetValue(logicalLocation, out int logicalLocationIndexFromMap)))
{
    logicalLocationIndex = logicalLocationIndexFromMap;
}
```

Reply:
Why I think it's clearer:
1. Doesn't require an uninitialized variable.
2. Doesn't use a temporary, and so...
3. One line shorter.
4. `if` statement uses "positive logic".

YMMV.

---
In reply to: [244239471](https://github.com/Microsoft/sarif-sdk/pull/1185#discussion_r244239471) [](ancestors = 244239471)

Reply:
I like where you're taking this. However, you neglect a key point, when TryGetValue fails, 'logicalLocationIndexFromMap' is initialized to the default of 0, which is not the actual default for logical location index (it is -1 as you know).

And so another subtlety in the format, our default values that do not correspond to C# primitive defaults require additional work in cases like this.

---
In reply to: [244239715](https://github.com/Microsoft/sarif-sdk/pull/1185#discussion_r244239715) [](ancestors = 244239715,244239471)

Reply:
Oh, right you are.

---
In reply to: [244372604](https://github.com/Microsoft/sarif-sdk/pull/1185#discussion_r244372604) [](ancestors = 244372604,244239715,244239471)